src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +994,746 @@ struct
   let to_yojson (x, y) = `Assoc [ (Base.name (), Base.to_yojson x); ("length", Idx.to_yojson y) ]
 end
 
-module AttributeConfiguredArrayDomain(Val: LatticeWithSmartOps) (Idx:IntDomain.Z):S with type value = Val.t and type idx = Idx.t =
+module NullByte (Val: LatticeWithNull) (Idx: IntDomain.Z): Str with type value = Val.t and type idx = Idx.t =
+struct
+  module MustNulls = SetDomain.Reverse (SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All Null" end))
+  module MayNulls = SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All Null" end)
+  (* (Must Null Set, May Null Set, Array Size) *)
+  include Lattice.Prod3 (MustNulls) (MayNulls) (Idx)
+
+  let name () = "arrays containing null bytes"
+  type idx = Idx.t
+  type value = Val.t
+
+  type ret = Null | NotNull | Top
+
+  (* helper: returns Idx.maximal except for Overflows that are mapped to None *)
+  let idx_maximal i = match Idx.maximal i with
+    | Some i -> (try Some (Z.of_int (Z.to_int i)) with Z.Overflow -> None)
+    | None -> None
+
+  let get ?(checkBounds=true) (ask: VDQ.t) (must_nulls_set, may_nulls_set, size) (e, i) =
+    let rec all_indexes_must_null i max =
+      if Z.gt i max then
+        true
+      else if MustNulls.mem i must_nulls_set then
+        all_indexes_must_null (Z.succ i) max
+      else
+        false in
+    let min interval = match Idx.minimal interval with
+      | Some min_num when Z.geq min_num Z.zero -> min_num
+      | _ -> Z.zero in (* assume worst case minimal natural number *)
+
+    let min_i = min i in
+    let max_i = idx_maximal i in
+    let min_size = min size in
+
+    (* warn if index is (potentially) out of bounds *)
+    if checkBounds then (array_oob_check (module Idx) ((must_nulls_set, may_nulls_set), size) (e, i));
+    match max_i, idx_maximal size with
```

Comment:
Is it maybe easier to understand if you pull all the cases where accesses are out of bounds and we thus return top out and the only deal with the well-defined cases in here?

Reply:
I personally don't think it makes things much easier to read, but I'm biased of course. Do you find this function difficult to understand in its current form?